ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -263,7 +263,11 @@ private MemoizedSupplier<List<CompletableFuture<Message>>> applyLog() throws Raf
         final long incremented = appliedIndex.incrementAndGet(debugIndexChange);
         Preconditions.assertTrue(incremented == nextIndex);
         if (f != null) {
-          futures.get().add(f);
+          applyLogFutures = applyLogFutures.thenCombine(f.exceptionally(ex -> {
+            LOG.error("Exception while {}: applying txn index={}, nextLog={}", this, nextIndex,
+                    LogProtoUtils.toLogEntryString(entry));
+            return null;
+          }), (v, message) -> null);
```

Comment:
Move out `f.exceptionally`, i.e.
```java
f.exceptionally(ex -> {
            LOG.error("Exception while {}: applying txn index={}, nextLog={}", this, nextIndex,
                    LogProtoUtils.toLogEntryString(entry));
            return null;
          })
applyLogFutures = applyLogFutures.thenCombine(f, (v, message) -> null);
```

Reply:
Both them are the same in our case.  Separating them makes it more readable.

Reply:
done